### PR TITLE
Attempt to reset rtl_tcp by directly channing frequency 

### PR DESCRIPTION
**Pull request by b0naf1de**
Commits in this pull request:

5572df3:  An attempt to prevent rtl_tcp from locking up and causing rtlamr to produce errors. Working well in my environment and should not impact other configurations. May need further testing in docker and spawned rtl_tcp configurations.

d779872: Adjusts sleep_for time based on time required to receive a full set of meter readings. In the previous model, the script would sleep for a predefined period after receiving the final meter reading in a set of meters.  With this commit, the sleep period receipt of the first meter reading in a set of meters. So, the time between meter readings is more deterministic.

### CODE REVIEW

**Comment by allangood** [comment]
Hi @b0naf1de , thank you for your contribution!

I have some reservation regarding using the sleep_for with a fixed period of time instead sleeping for after a successful reading.
My user case for instance would probably break with a fixed time. My meter for instance, sends readings on random period of times. This change can potentially read only when my meter is sleeping.

Also, the `rtlamr` has a `-duration=` parameter that can be used if you want some predictability.
https://github.com/bemasher/rtlamr/wiki/Configuration

Would you mind to remove the d779872 from your PR?

Thank you very much!

**Comment by b0naf1de** [comment]
I can definitely hold it and further test the sleep process. But, I don't think I am introducing a fixed sleep period; the fixed sleep period is already in rtlamr2mqtt whenever sleep_for is >0.  Can you help me understand what you mean by sleep_for with fixed period of time?

Apologies for TL;DR or if I am missing your point.
In the current implementation, rtlamr2mqtt sleeps for a fixed duration after a successful reading. Actually, it sleeps for a fixed duration after **at least one successful reading** from **every** meter. The fixed time begins after a message is received from the last meter.  This creates a problem when multiple meters are configured. Here's an example, simplified to use minutes.

```
sleep_for = 3
Meter A:  00:00:00 msg A1
Meter A:  00:05:00 msg A2
Meter B:  00:07:00 msg B1 --> we now have >0 readings from each meter so sleep_for 3 minutes
Meter A:  00:09:00 message missed because we're sleeping
          00:10:00 --> wakeup
```

The  [d779872](https://github.com/allangood/rtlamr2mqtt/commit/d7798727550c525bcccb31b9e10d33ab51098e2e) patch reduces the sleep_time to avoid missing messages. Instead of using a fixed sleep_for period that begins after when a message is received from the last meter, it reduces sleep_for based on a message is received from the first meter.  
 
```
sleep_for = 3
Meter A:  00:00:00 msg A1
Meter A:  00:05:00 msg A2
Meter B:  00:07:00 msg B1 --> we now have >0 readings from each meter. Only sleep for 1 minute because 
                              2 minutes have elapsed since most recent Meter A message was received
          00:08:00 --> wake up
Meter A:  00:09:00 msg A1 --> message received
```

So, I think the patch causes less sleep in all cases, never move.

Also, an observation that may help others. The r900 protocol frequency hops ([every 14 seconds](https://sandeen.net/wordpress/energy/neptune-rf-water-meter-frequency-hopping-pattern/)). With the default 32 or 72 symbol length, the radio will pickup multiple bins of transmission, making it look like the period between messages is random. On my meter, within a 15 minute window I will receive and successfully decode multiple messages for the meter. However, the actual consumption value is only updated every 15 minutes. I tested this by:  (1) recording consumption, (2) flushing toilet, (3) recording consumption reading until I see consumption change (4) flushing toilet again, (5) recording more readings until consumption changes again. The time between (3) and (5) is never less than 15 minutes even though my meter reported several messages during this period. So, the period that the value is updated is independent from how often messages are sent. This may vary for different meters but is a decent way to find your maximum sleep time.

**Comment by allangood** [comment]
Hi @b0naf1de ,

Now I get it what you wanted! Sorry, I didn't have enough time to look your code! :)
Awesome ELI5 explanation and even better tests! I really appreciate all your effort and time! Thank you very much!

What I tried to say with the "sleeping for a fixed period of time" was to have the container finishing all process after a `sleep_for` period of time, with or without a reading. I said that because I misunderstood your code.

TL;DR; `sleep_for` is an optional parameter and missing readings is acceptable if you choose to use it. You have to weight what is more important: CPU usage or Meter readings. If you don't want to miss a reading then `sleep_for: 0` should be used.

My thoughts on this are:

1. `sleep_for` was created to use less resources and it is an optional parameter
2. If loosing messages is something that an user case cannot afford, then `sleep_for` shouldn't be used in that case
3. Using your examples, for my user case, having the container sleeping for 5 minutes is more important than catching the second message. My meter sends a lot of repeated messages, losing some are not really a big deal. In the other hand, if the container doesn't sleep for the whole 5 minutes, my RPi doesn't have enough time to cool down and then runs too hot and crashes. If or when this happens, I have to shutdown my RPi, wait it until it cools down and the start it gain. Not fun :)
4. Even in your example no.2, it is possible to have some missing readings, but IMO, it really doesn't matter as it can catch it up latter.

What are your thoughts?

**Comment by b0naf1de** [comment]
Alright, let me think about how to rework that patch.  What I'm going for is something that ensures that I get at least one reading within every 15 minute period while allowing as much sleep as possible. If you have having thermal challenges I can see where -duration= and sleep_for combination would ensure you give your system a chance to cool down. I've removed the commit from the pull request.

**Comment by allangood** [comment]
Thank you very much for your contribution! Everything looks awesome! :)
